### ISSUE TITLE
Points reactor: don't refire when a post's status is toggled

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#53]
## The Problem

We need to make it so that reactors don't fire multiple times for the same event. This actually depends on the reactor. If the reactor listens for reversals and spam, then there is no need to have special handling for when a reversal and then a refire occurs. For the badges component, for example, we don't need to worry about this, because we'll be using the retroactive API. But for the points component we need to make sure we don't award points multiple times when, for example, a user toggles a post's status multiple times.
## History

In the past what we have done is check if we had already awarded points to that post by querying the points logs. If we had, we'd skip awarding the points. For comments, we did things a little differently, saving a piece of metadata for the comment that told what the last status of that comment was. Then we'd know whether to add points (if the comment was changing from unapproved to approved) or not (if the status was the same).

The reason for the discrepancy between the way that posts and comments were handles is that in the points hooks API the post points hook did not reverse except when a post was deleted, instead of when it was unpublished. So we only ever awarded once per post. The comments hook on the other hand would reverse whenever the comment was unapproved, so that we might award and then subtract points several times as a comment's status was toggled.

In the new API we decided to go with only listening for spam in the points reactor instead of all reversals, see #2. So at the moment there are no reversals for points awarded for posts, since there is no spam status for posts. (Note that at the moment the reverse action for the post publish event is still post delete, but this may change before the API is finalized. It doesn't really matter for the current discussion anyway; like I said, the spam action is listened for, not reverse.) Points are removed for comments when they are marked as spam though.

The issue is that points are also awarded again when a post's or comment's status is toggled multiple times, and as long as it isn't toggled to the spam status (which there is none for posts by default), no points will be removed.
### Metadata

If we wanted to, we could do something similar to what the points hooks API did. We could set a meta flag or run a query to see if points were already awarded. The problem with setting a flag is that we don't have any clue what sort of entities we're dealing with, so we can't just assume there will be a metadata API for each one, and even if there was we'd have no way of knowing what it was/how to use it.
### Logs Queries

That leaves the option of querying the logs, but there are problems with that as well. First, it can result in more expensive queries, which definitely isn't ideal. Also, when we have something like the comment event which might toggle several times, it isn't possible to just check if any logs exist at all, we have to know whether there are more awards than reversals or not. Although since we're just listening to spam right now, that isn't a problem, because when a spam event occurs the logs are deleted. So at the moment I guess we can just run a logs query to determine this. Also, we could still add reversal to the points target later and we could add the other features necessary to handle that then.
### Reversals Log

Another option would be to listen to reversals and then log them somehow (in their own table I guess). We'd need to log the hits too. Then we'd be able to count the number of hits and the number of reversals, and tally up the results to see if all hits have been reversed.

This seems like a strange thing to do just for the sake of this though. It would mean that we could optimize the queries, but we'd still be needing to count the reversals and fires, instead of just pulling a single row from the database. Of course, I suppose that we could just save a single row for each event occurrence, and then when a reversal occurs just delete the row if it exists instead of adding a new row. That could give us the best of both worlds, but it would mean another database table that could grow rather large.

**Comment by JDGrimes** [comment]
I think that a reversals log might make sense if we one day start actually listening to reversals for this reactor, but until then the best solution might be running a points log query.

**Comment by JDGrimes** [comment]
## Importing existing hits

Note that whatever we do, we'll need to take into account that the points hooks API will have already awarded points for some past events. So we'll either need to keep checking the old method of detecting this for back-compat on legacy sites, or else import the old flags when we upgrade.

**Comment by JDGrimes** [comment]
## Not all events

It has just occurred to me that not all events can run only once per arg value. Some, like the User Visit event, are meant to occur multiple times. So we'd want to allow some events to repeat, but not others.

How would we detect which events are repeatable and which are not? We could check if any `reverse` or `spam` actions are registered for an event, and if there are none, assume that it is repeatable. But I'm not sure we should chain repeatability to reversibility—it's entirely possible that we'll want to have non-repeatable events that really aren't reversible either. I suppose that it's even possible that some repeatable events will also be reversible.

So really the only way that we can detect this is if we introduce a `WordPoints_Hook_Event_Repeatable` and implement it in the User Visit event and any other repeatable events.

**Comment by JDGrimes** [comment]
## Only points reactor?

> We need to make it so that reactors don't fire multiple times for the same event. This actually depends on the reactor. If the reactor listens for reversals and spam, then there is no need to have special handling for when a reversal and then a refire occurs. For the badges component, for example, we don't need to worry about this, because we'll be using the retroactive API. But for the points component we need to make sure we don't award points multiple times when, for example, a user toggles a post's status multiple times.

It is not local only the points reactor, but for all reactors that don't handle reversals as necessary to avoid repeats. So it makes sense to develop this as a generic piece of API. However, how we do that will depend upon whether there might ever be a reactor that would specifically want to handle repeats in a different manner than as would occur if we made this a part of the basic API. To allow for that, we could confine the repeat handling to a child of the reactor class, which reactors could extend if they wanted to handle repeatable events this way.

**Comment by JDGrimes** [comment]
## Just the Post Publish event

It actually turns out that the post status toggling is a special case, and needs special handling beyond what would normally be necessary to avoid rehits form an event.

The post publish event is hooked to the `post_publish` action to fire it, but to the `post_delete` action to reverse it. This means that the post being unpublished (returned to draft status, for example) will not cause it to be reversed. This allows for the event to be fired multiple times without being reversed, which wouldn't be possible if the event was reversed when the post was "unpublished".

The comment leave event, for example, does not suffer from this problem, because it is reversed with the comment is deapproved. This means that the event can't fire twice without being reversed, because the comment approve action can't fire again until the comment deapproved action fires (which will, of course, cause the hit to be reversed).

So the problem here is really that the post publish event is not configured properly. The correct thing to do is to hook it to post unpublish for reversal instead of post delete. As it is, the event is basically a repeatable event, because it isn't coupled tightly to the entity status changes.

That said, we need to consider maintaining back-compat here, for the points hooks that we'll import from the old API. Because of that, we'll probably need a two-pronged solution: we'll want to reverse the event on post unpublish, but we'll also need an alternative version of the event that has special refire handling—unless we decide that it is OK to change the behavior of the old points hooks after import.

**Comment by JDGrimes** [comment]
Even after changing the action the event reverses on I was still seeing issues with a test that I was running on this. I was toggling the status of a post several times and checking the number of points that the target user had. For some reason the user wasn't having their points removed on reverse. I found out that the reverse action was being called correctly, but that the points reactor wasn't finding the logs to reverse because the slash was being stripped from `post\post` before it was saved as a key in the meta table. So the value in the DB was `postpost`. So the query wasn't returning any results. I guess the slashing monster has come to bite us. We can tame it, but that's for another issue.

**Comment by JDGrimes** [comment]
Slashing in metadata API is a total nightmare, and still even has bugs: https://github.com/WordPoints/wordpoints/issues/364#issuecomment-182414973. We basically have to decide whether we want to:
- Add a workaround in WordPoints for a bug in `update_metadata()` (see above linked comment).
- Choose a different delimiter other than the backslash in our arg slugs.
- Stop saving the metadata this way, or use different keys.
- Or, possibly some combination of the above.

We were actually considering doing that third one anyway (#80), so we might want to explore that first. Although I'm sure that backslashes may come back to haunt us elsewhere later, so we might still want to consider changing the delimiter as well. But that would be for another ticket.

**Comment by JDGrimes** [comment]
I'm going to stick with backslashes as the delimiter for now, but we can revisit in another ticket if needed. We are planning to work around that bug in WordPoints (https://github.com/WordPoints/wordpoints/issues/366). Relying on WordPoints handling slashes in meta keys (even without the workaround for that bug in WordPress), means that we're no longer compatible with WordPoints stable, only the dev version. I'm not sure that was ever really intended anyway, but we need to update our Travis CI config to reflect it.

**Comment by JDGrimes** [comment]
This is basically fixed by #94. We may still want to pursue #66, but at the moment it is a separate issue from this.

This ticket was based on the assumption that we needed special handling to avoid rehits when the status of a post was toggled, but that is not actually true. Reversals will automatically occur when the status is toggled back (now that we're using the correct reverse action for the event). Whether we add special handling beyond that (like for the extreme edge-case that the action isn't called when the status is toggled in some non-standard way) is a separate issue.
